HW2/P3/mandelbrot.pyx

```diff
@@ -30,18 +30,61 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
     with nogil:
-        for i in range(in_coords.shape[0]):
-            for j in range(in_coords.shape[1]):
-                c = in_coords[i, j]
-                z = 0
+        #for i in range(in_coords.shape[0]):
+        for i in prange(in_coords.shape[0], schedule="static", 
+                        chunksize=1, num_threads=1):
+            for j in range(0,in_coords.shape[1],8):
+                avx_real = AVX.make_float8(in_coords[i, j+7].real,
+                                           in_coords[i, j+6].real,
+                                           in_coords[i, j+5].real,
+                                           in_coords[i, j+4].real,
+                                           in_coords[i, j+3].real,
+                                           in_coords[i, j+2].real,
+                                           in_coords[i, j+1].real,                  
+                                           in_coords[i, j].real
+                                          )
+                avx_imag = AVX.make_float8(in_coords[i, j+7].imag,
+                                           in_coords[i, j+6].imag,
+                                           in_coords[i, j+5].imag,
+                                           in_coords[i, j+4].imag,
+                                           in_coords[i, j+3].imag,
+                                           in_coords[i, j+2].imag,
+                                           in_coords[i, j+1].imag,           
+                                           in_coords[i, j].imag
+                                          )
+                avx_zreal = AVX.float_to_float8(0.0)
+                avx_zimag = AVX.float_to_float8(0.0)
+                avx_iter = AVX.float_to_float8(0.0)
                 for iter in range(max_iterations):
-                    if magnitude_squared(z) > 4:
-                        break
-                    z = z * z + c
-                out_counts[i, j] = iter
-
+                    avx_mag = AVX.add(AVX.mul(avx_zreal, avx_zreal), 
+                                     AVX.mul(avx_zimag, avx_zimag) )
+                    # mask will be true where 4.0 < avx_mag                
+                    avx_mask = AVX.greater_than(AVX.float_to_float8(4.0), avx_mag)
+                    avx_zsqre_real = AVX.sub(AVX.mul(avx_zreal, avx_zreal), 
+                                AVX.mul(avx_zimag, avx_zimag) )
+                    avx_zsqre_imag = AVX.mul(AVX.mul(avx_zreal, avx_zimag), 
+                                             AVX.float_to_float8(2.0))
+                    avx_zreal = AVX.add(avx_zsqre_real, avx_real)
+                    avx_zimag = AVX.add(avx_zsqre_imag, avx_imag)
+                    avx_mask_iter = AVX.bitwise_and(AVX.float_to_float8(1),
+                                                avx_mask
+                                                )
+                    avx_iter = AVX.add(avx_iter, avx_mask_iter)
+                AVX.to_mem(avx_iter, &(out_vals[0]))
```

Comment:
Interesting approach, executing the to_mem call here. Did you experience any problems with multiple threads trying to execute this concurrently?